wagtail/images/views/serve.py

```diff
@@ -35,50 +60,72 @@ def as_view(cls, **initkwargs):
+        self,
+        request: "HttpRequest",
+        signature: str,
+        image_id: int,
+        filter_spec: str,
+        filename: Optional[str] = None,
+    ):
         if not verify_signature(
             signature.encode(), image_id, filter_spec, key=self.key
         ):
-            raise PermissionDenied
+            return self.get_error_response("Invalid signature", HTTPStatus.BAD_REQUEST)
 
-        image = get_object_or_404(self.model, id=image_id)
+        image = self.get_image(image_id)
 
         # Get/generate the rendition
         try:
             rendition = image.get_rendition(filter_spec)
         except SourceImageIOError:
-            return HttpResponse(
-                "Source image file not found", content_type="text/plain", status=410
+            return self.get_error_response(
+                "Source image file not found", HTTPStatus.GONE
             )
         except InvalidFilterSpecError:
-            return HttpResponse(
-                "Invalid filter spec: " + filter_spec,
-                content_type="text/plain",
-                status=400,
+            return self.get_error_response(
+                f"Invalid filter spec: {filter_spec}", HTTPStatus.BAD_REQUEST
             )
+        # Make the image object available without an additional query
+        rendition.image = image
+
+        return self.get_success_response(rendition)
 
+    def get_image(self, image_id: int) -> "AbstractImage":
+        return get_object_or_404(self.model, id=image_id)
+
+    def get_success_response(
+        self, rendition: "AbstractRendition"
+    ) -> "HttpResponseBase":
         return getattr(self, self.action)(rendition)
 
-    def serve(self, rendition):
-        with rendition.get_willow_image() as willow_image:
-            mime_type = willow_image.mime_type
+    def get_error_response(self, message: str, status: HTTPStatus) -> HttpResponse:
+        response = HttpResponse(message, content_type="text/plain", status=status)
+        if self.error_cache_control_headers:
+            patch_cache_control(response, **self.error_cache_control_headers)
+        return response
 
+    def serve(self, rendition: "AbstractRendition") -> "FileResponse":
         # Serve the file
         rendition.file.open("rb")
-        response = FileResponse(rendition.file, content_type=mime_type)
+        response = FileResponse(rendition.file, content_type=rendition.get_mime_type())
 
         # Add a CSP header to prevent inline execution
         response["Content-Security-Policy"] = "default-src 'none'"
 
         # Prevent browsers from auto-detecting the content-type of a document
         response["X-Content-Type-Options"] = "nosniff"
 
+        if self.serve_cache_control_headers:
+            patch_cache_control(response, **self.serve_cache_control_headers)
         return response
 
-    def redirect(self, rendition):
+    def redirect(self, rendition: "AbstractRendition") -> "HttpResponseRedirectBase":
```

Comment:
NOTE: Using `HttpResponseRedirectBase` here, because I think it's perfectly reasonable to want to return a permanent redirect under some circumstances, and there's no need for type-checkers to get in the way of that.